File: First-Edition/BookSamples/rosetta-stone/rust/rsc1p3_type_level_interfaces/src/cellular/language/extensions.rs
use crate::automaton::IState;
use crate::automaton::INeighborhood;
use crate::automaton::IStep;
use crate::automaton::ICellCondition;
use crate::automaton::IRule;

use std::marker::PhantomData;
use type_level::IInterface;
use type_level::Wrapper;
use tl_list_lib::I32List;
use tl_str_list::TlStr;


pub struct NeighborsCountImpl<S, Cnts: I32List>
  (PhantomData::<(S, Cnts)>);

pub type NeighborsCount<S, Cnts> =
  Wrapper<ICellCondition, NeighborsCountImpl<S, Cnts>>;


pub struct AdjacentsLvlImpl<const LVL: u8>;

pub type AdjacentsLvl<const LVL: u8> =
  Wrapper<INeighborhood, AdjacentsLvlImpl<LVL>>;


pub struct StateImpl<Name: TlStr, const IDX: u8> (PhantomData::<Name>);

pub type State<Name, const IDX: u8> =
  Wrapper<IState, StateImpl<Name, IDX>>;


pub struct RuleImpl <
    Name: TlStr,
    Code: TlStr,
    Neighborhood: IInterface<INeighborhood>,
    Step: IInterface<IStep>>
      (PhantomData::<(Name, Code, Neighborhood, Step)>);

pub type Rule<N, C, Nh, S> =
  Wrapper<IRule, RuleImpl<N, C, Nh, S>>;

File: First-Edition/BookSamples/rosetta-stone/scala3/rsc2p4_correctness_validation/src/main/scala/cellular/Extensions.scala
package cellular
package extensions

import graninas.typelevel._
import cellular.language._


case class Literal[M]()


// Parameterized implementation type, option 1: type parameters

case class StateImpl[
  Name <: String & Singleton,
  Idx <: Int & Singleton]()

type State[Name <: String & Singleton, I <: Int & Singleton] =
  MkState[StateImpl[Name, I]]


// Parameterized implementation type, option 2: associated types
// TODO: double-check how to interpret it
sealed trait NeighborsCountImpl:
  type State <: IState
  type Counts <: IntList

type NeighborsCount[S <: IState, Cnts <: IntList] =
  MkCellCondition[NeighborsCountImpl {
    type State = S
    type Counts = Cnts
  }]


case class StateTransitionImpl[
  From <: IState,
  To <: IState,
  Conds <: ICellCondition]()

type StateTransition[
  From <: IState,
  To <: IState,
  Conds <: ICellCondition] =
  MkStateTransition[StateTransitionImpl[From, To, Conds]]


case class AdjacentsLvlImpl[
  Lvl <: Int & Singleton]()

type AdjacentsLvl[Lvl <: Int & Singleton] =
  MkNeighborhood[AdjacentsLvlImpl[Lvl]]


case class StepImpl[
  DefState <: IState,
  Transitions <: HList[IStateTransition]]()

type Step[
  DefState <: IState,
  Transitions <: HList[IStateTransition]] =
  MkStep[StepImpl[DefState, Transitions]]


case class RuleImpl[
  Name <: String & Singleton,
  Code <: String & Singleton,
  Neighborhood <: INeighborhood,
  Step <: IStep]()

type Rule[
  Name <: String & Singleton,
  Code <: String & Singleton,
  Neighborhood <: INeighborhood,
  Step <: IStep] =
  MkRule[RuleImpl[Name, Code, Neighborhood, Step]]
